### PR TITLE
between adapter

### PR DESCRIPTION
**Pull request by digitalist**
fixing test coverage! :)

### CODE REVIEW

File: adapters/between/betweentest/exemplary/simple-site-banner.json
Comment (mansinahar) [inline adapters/between/betweentest/exemplary/simple-site-banner.json:2]: Could you please also add another test case that has multiple imps where a subset of imps might have the same bidder ext. This to exercise the `splitImpressions` code

File: adapters/between/betweentest/params/race/banner.json
Comment (mansinahar) [inline adapters/between/betweentest/params/race/banner.json:2]: This file looks like a copy-paste leftover. The params here should be the params that you expect in the between bidder ext which for you is just `host`

File: adapters/between/usersync_test.go
Comment (mansinahar) [inline adapters/between/usersync_test.go:8]: From the rest of the code it looks like your adapter doesn't support usersyncing and so can you please remove this file and the `usersync.go` file? Thanks

File: usersync/usersyncers/syncer.go
Comment (mansinahar) [inline usersync/usersyncers/syncer.go:152]: Since your adapter doesn't support usersyncing, you don't need this

**Comment by digitalist** [comment]
thanks for the code review, will fix soon

File: usersync/usersyncers/syncer.go
Comment (digitalist) [reply usersync/usersyncers/syncer.go:152]: I'll delete all these usersync leftovers for now, and return later, after we done with usersync testing internally, thanks for pointing this out.

Comment (mansinahar) [reply usersync/usersyncers/syncer.go:152]: Sounds great! Thanks @digitalist

File: adapters/between/betweentest/exemplary/multi-request.json
Comment (mansinahar) [inline adapters/between/betweentest/exemplary/multi-request.json:9]: I'd suggest adding another imp here with different `host` value as that is how you're clubbing the imps together for each request to your adapter

**Comment by mansinahar** [comment]
@digitalist Can you please open up a PR for your adapter in the [docs](https://github.com/prebid/prebid.github.io/tree/master/dev-docs/bidders) repo too? Prebid.org has rolled out these new docs that requires docs for each adapter

**Comment by SyntaxNode** [comment]
@digitalist Friendly ping :). No rush if you're busy on other things at the moment.

**Comment by digitalist** [comment]
we're just thinking about what we want to implement right now.

File: adapters/between/between.go
Comment (digitalist) [reply adapters/between/between.go:123]: guess fixing it with a `continue` instead of returning an error would be fine?

File: adapters/between/between.go
Comment (digitalist) [reply adapters/between/between.go:51]: done, I think this will be not a sourceId, but a publisherId AND sourceId/zoneId if all goes well

**Comment by SyntaxNode** [comment]
We'll take a look soon. Please remember to open a documentation PR for the https://github.com/prebid/prebid.github.io repo following the instructions here https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html#document-your-adapter. We can't merge this PR in without documentation.

File: adapters/between/between.go
Comment (mansinahar) [inline adapters/between/between.go:131]: Don't you still need this line here? If you do't append the imp to the map value, you'll always only have one (the last) imp for a give set of bidder params

File: adapters/between/between.go
Comment (mansinahar) [inline adapters/between/between.go:125]: I'd recommend having this function return an error slice and append the error here to that slice before calling `continue` and then return that slice at the end

File: adapters/between/betweentest/exemplary/multi-request.json
Comment (mansinahar) [inline adapters/between/betweentest/exemplary/multi-request.json:57]: Shouldn't the expected request to your service have both the impressions in it?

File: adapters/between/between.go
Comment (mansinahar) [reply adapters/between/between.go:74]: I think this comment was hidden so responding on it for visibility

File: adapters/between/between.go
Comment (mansinahar) [reply adapters/between/between.go:102]: Same as the previous comment. Responding on it for visibility

File: adapters/between/between.go
Comment (mansinahar) [reply adapters/between/between.go:169]: Same as the previous comment. Responding on it for visibility

**Comment by mansinahar** [comment]
@digitalist Can you please also scroll up and click on `Hidden conversation`? There are three comments there that I believe went unnoticed because they were hiding there. I tried to respond on those now to push them down and make them visible but it didn't work unfortunately :(

File: adapters/between/between_test.go
Comment (guscarreon) [inline adapters/between/between_test.go:10]: Is this comment necessary? Would you consider removing it?

**Comment by SyntaxNode** [comment]
FYI: Corresponds to https://github.com/prebid/prebid.github.io/pull/2299

File: adapters/between/betweentest/exemplary/multi-request.json
Comment (digitalist) [reply adapters/between/betweentest/exemplary/multi-request.json:57]: done, pushing soon

File: adapters/between/between.go
Comment (digitalist) [reply adapters/between/between.go:125]: done, pushing soon

Comment (digitalist) [reply adapters/between/between.go:125]: thanks for the tip, by the way

File: adapters/between/between.go
Comment (digitalist) [reply adapters/between/between.go:131]: done, pushing soon

File: adapters/between/between_test.go
Comment (digitalist) [reply adapters/between/between_test.go:10]: removed, thanks

File: adapters/between/between.go
Comment (digitalist) [reply adapters/between/between.go:102]: done, pushing soon

File: adapters/between/between.go
Comment (digitalist) [reply adapters/between/between.go:169]: done:-)

File: adapters/between/between.go
Comment (digitalist) [reply adapters/between/between.go:32]: whoops, I see an error there, fixing r/n

File: adapters/between/betweentest/exemplary/multi-request.json
Comment (digitalist) [reply adapters/between/betweentest/exemplary/multi-request.json]: I'm in doubt we'll have different hosts in an impressions near future

File: adapters/between/between.go
Comment (digitalist) [reply adapters/between/between.go:117]: fixed

File: adapters/between/between.go
Comment (digitalist) [reply adapters/between/between.go]: fixed

**Review by mansinahar — DISMISSED** [review]

**Comment by mansinahar** [comment]
Thanks @digitalist for making the requested changes. LGTM!

File: config/config.go
Comment (guscarreon) [inline config/config.go:971]: Can we move this entry to its alphabetical order below line 912?

File: openrtb_ext/bidders.go
Comment (guscarreon) [inline openrtb_ext/bidders.go:106]: Can we keep this list alphabetically ordered? Can we move entry below line 47?

File: openrtb_ext/bidders.go
Comment (guscarreon) [inline openrtb_ext/bidders.go:191]: Can this entry be moved to its alphabetical order spot above?

**Review by guscarreon — CHANGES_REQUESTED** [review]
The adapter looks great. Before approving, can we expand the test coverage to at least 90%?
```
╰─➤  go test -coverprofile=c.out
PASS
coverage: 81.2% of statements
╰─➤  go tool cover -func=c.out
github.com/prebid/prebid-server/adapters/between/between.go:21:         MakeRequests            80.0%
github.com/prebid/prebid-server/adapters/between/between.go:69:         MakeBids                93.8%
github.com/prebid/prebid-server/adapters/between/between.go:111:        splitImpressions        83.3%
github.com/prebid/prebid-server/adapters/between/between.go:132:        getBidderParams         66.7%
github.com/prebid/prebid-server/adapters/between/between.go:156:        getMediaType            100.0%
github.com/prebid/prebid-server/adapters/between/between.go:163:        NewBetweenBidder        60.0%
total:                                                                  (statements)            81.2%
```

File: openrtb_ext/bidders.go
Comment (digitalist) [reply openrtb_ext/bidders.go:106]: done, pushing soon.

File: openrtb_ext/bidders.go
Comment (digitalist) [reply openrtb_ext/bidders.go]: fixed

File: config/config.go
Comment (digitalist) [reply config/config.go]: fixed

**Review by guscarreon — APPROVED** [review]
@digitalist thank you for addressing the feedback. Adapter looks great.

File: config/config.go
Comment (SyntaxNode) [inline config/config.go:913]: Why do you need publishers to provide the domain? PBS does not act as a general proxy. I realize a few other adapters do this today, but it's something we're still working with bidders to fix.

Comment (SyntaxNode) [reply config/config.go:913]: I see you marked this resolved, but I didn't see your answer. I might have missed a reply in another comment thread.

**Review by digitalist — COMMENTED** [review]
>that's our business logic so far to handle geo-related issues. 
maybe we'll change this in a near feature but as of now that's what we need to to have as a parameter.

**Comment by mansinahar** [comment]
@digitalist Are all these hosts going to have a different domain or only a certain part of the endpoint will change from publisher to publisher? If it's the latter than maybe you can add a macro for the part that would change and have the common part hardcoded in the default endpoint?

**Comment by digitalist** [comment]
> @digitalist Are all these hosts going to have a different domain or only a certain part of the endpoint will change from publisher to publisher? If it's the latter than maybe you can add a macro for the part that would change and have the common part hardcoded in the default endpoint?

I guess our server team wants to have as much freedom to change/setup hosts as possible as of now. I'm sure your macros idea is what I'll have to patch firstly, but as of know I would like to merge it as it is.

**Comment by SyntaxNode** [comment]
> I guess our server team wants to have as much freedom to change/setup hosts as possible as of now. I'm sure your macros idea is what I'll have to patch firstly, but as of now I would like to merge it as it is, if that's possible.

We discussed your request at the Prebid Server Committee (a cross-company PrebigOrg working group) and we will not permit new adapters to allow the publisher to specify a full host name. This is to ensure hosts have sufficient control over their outgoing connections so they may maintain a stable hosting environment. This also prevents Prebid Server from acting as a generic proxy.

Parts of a host name which differ due to publisher ids or geographical regions (your use case) are permitted, provided that the rest of the domain is account for in the configured endpoint url. I would also welcome a small list of known host names that the adapter validates.

The Prebid Server Committee will be reaching out to the current adapters which do permit the publisher to specify a full host name to resolve current issues. We will also update the adapter writing guide to clarify this rule and reasoning.

I hope we can reach a compromise with your server team so that we can proceed and merge in your adapter.

**Review by mansinahar — APPROVED** [review]
LGTM! Just wanna make sure that the `/s2s` in the endpoint path was removed intentionally as you had that earlier

**Comment by digitalist** [comment]
> LGTM! Just wanna make sure that the `/s2s` in the endpoint path was removed intentionally as you had that earlier

yes, I guess we can parametrize by hostnames, so it's ok, thank you

**Comment by mansinahar** [comment]
Perfect! Thanks for the response. Once we have another approval, we can go ahead and merge

**Review by guscarreon — APPROVED** [review]
LGTM

**Pull request merged**
